File: sdk/react-native/ios/SentoriCrashHandler.swift
import Foundation

/// Static crash handler — captures NSException and writes one JSON file
/// per crash to <Documents>/sentori/pending/<uuid>.json. JS drains that
/// directory on next launch via `Sentori.drainPending()`.
///
/// What this does NOT do (Phase 7 v0.1):
///   - signal-based native crashes (SIGSEGV / SIGABRT etc.) — see ROADMAP
///     "explicitly out" list. Only Objective-C exceptions are caught here.
@objc public final class SentoriCrashHandler: NSObject {

    private static let configKey = "com.sentori.config"
    private static let pendingDirName = "sentori/pending"

    /// Install the global uncaught-exception handler. C function pointer,
    /// so we cannot capture local context (no chaining to a previously
    /// installed handler in v0.1 — RedBox in dev still receives via
    /// JS-side handlers; in release this just replaces the default).
    @objc public static func register() {
        NSSetUncaughtExceptionHandler(SentoriCrashHandler.exceptionHandler)
    }

    private static let exceptionHandler: @convention(c) (NSException) -> Void = { exception in
        SentoriCrashHandler.write(exception: exception)
    }

    /// JS side calls this on `sentori.init(...)` so the crash handler
    /// has release / environment when an exception fires later.
    @objc public static func setConfig(_ config: [String: Any]) {
        UserDefaults.standard.set(config, forKey: configKey)
    }

    /// Read all pending-crash files, return their contents (UTF-8 JSON
    /// strings), and remove them from disk. Best-effort: any I/O error
    /// drops that one file silently.
    @objc public static func consumePending() -> [String] {
        guard let dir = pendingDir() else { return [] }
        let urls = (try? FileManager.default.contentsOfDirectory(
            at: dir, includingPropertiesForKeys: nil)) ?? []
        var out: [String] = []
        for url in urls where url.pathExtension == "json" {
            if let data = try? Data(contentsOf: url),
               let str = String(data: data, encoding: .utf8) {
                out.append(str)
            }
            try? FileManager.default.removeItem(at: url)
        }
        return out
    }

    // MARK: - Internals

    private static func pendingDir() -> URL? {
        guard let docs = FileManager.default.urls(
            for: .documentDirectory, in: .userDomainMask).first else { return nil }
        let dir = docs.appendingPathComponent(pendingDirName)
        try? FileManager.default.createDirectory(
            at: dir, withIntermediateDirectories: true)
        return dir
    }

    private static func config() -> [String: Any] {
        return UserDefaults.standard.dictionary(forKey: configKey) ?? [:]
    }

    private static func write(exception: NSException) {
        let cfg = config()
        let release = (cfg["release"] as? String) ?? "unknown"
        let environment = (cfg["environment"] as? String) ?? "prod"

        var event: [String: Any] = [
            "id": UUID().uuidString.lowercased(),
            "timestamp": iso8601(Date()),
            "kind": "error",
            "platform": "ios",
            "release": release,
            "environment": environment,
            "device": [
                "os": "ios",
                "osVersion": osVersion(),
                "model": deviceModel(),
            ],
            "app": appInfo(),
            "user": NSNull(),
            "tags": [String: String](),
            "breadcrumbs": [Any](),
            "error": [
                "type": exception.name.rawValue,
                "message": exception.reason ?? "",
                "stack": frames(from: exception.callStackSymbols),
                "cause": NSNull(),
            ],
            "fingerprint": [String](),
            "traceId": NSNull(),
            "spanId": NSNull(),
        ]

        // Phase 42 sub-E.04/07: capture the screen + view tree right
        // before the app dies. The handler runs synchronously on the
        // thread that threw NSException; UIKit's still valid at this
        // point. Both blobs go in a temp `_pendingAttachments` field
        // — JS strips it on next launch, uploads each via
        // `POST /v1/events/<id>/attachments/<kind>`, then enqueues
        // the cleaned event.
        if let snap = SentoriScreenshotCapture.captureKeyWindow() {
            var pending: [[String: Any]] = []
            if let sc = snap["screenshot"] as? [String: Any],
               let b64 = sc["base64"] as? String {
                pending.append([
                    "kind": "screenshot",
                    "base64": b64,
                    "mediaType": (sc["mediaType"] as? String) ?? "image/jpeg",
                    "source": "ios",
                ])
            }
            if let vt = snap["viewTree"],
               let data = try? JSONSerialization.data(withJSONObject: vt, options: []) {
                pending.append([
                    "kind": "viewTree",
                    "base64": data.base64EncodedString(),
                    "mediaType": "application/json",
                    "source": "ios",
                ])
            }
            if !pending.isEmpty {
                event["_pendingAttachments"] = pending
            }
        }

        guard let dir = pendingDir() else { return }
        let url = dir.appendingPathComponent("\(UUID().uuidString.lowercased()).json")
        if let data = try? JSONSerialization.data(withJSONObject: event, options: []) {
            try? data.write(to: url)
        }
    }

    private static func iso8601(_ date: Date) -> String {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f.string(from: date)
    }

    private static func osVersion() -> String {
        let v = ProcessInfo.processInfo.operatingSystemVersion
        return "\(v.majorVersion).\(v.minorVersion).\(v.patchVersion)"
    }

    private static func deviceModel() -> String {
        var s = utsname()
        uname(&s)
        return withUnsafePointer(to: &s.machine) { ptr in
            ptr.withMemoryRebound(to: CChar.self, capacity: 1) {
                String(validatingUTF8: $0) ?? "unknown"
            }
        }
    }

    private static func appInfo() -> [String: Any] {
        let info = Bundle.main.infoDictionary ?? [:]
        var d: [String: Any] = [
            "version": (info["CFBundleShortVersionString"] as? String) ?? "0.0.0",
        ]
        if let build = info["CFBundleVersion"] as? String {
            d["build"] = build
        }
        return d
    }

    /// Best-effort frame parse from `[NSException callStackSymbols]`.
    /// Each line looks roughly like:
    ///   "1   AppName    0x0001a0b0 -[ClassName method:] + 100"
    /// We don't have file/line info in raw symbol output; sourcemap-style
    /// symbolication for native happens server-side (Phase 8+).
    private static func frames(from symbols: [String]) -> [[String: Any]] {
        return symbols.map { sym -> [String: Any] in
            let parts = sym.split(separator: " ", omittingEmptySubsequences: true).map(String.init)
            let module = parts.count > 1 ? parts[1] : "<unknown>"
            let function = parts.count > 3 ? parts.dropFirst(3).joined(separator: " ") : "<anonymous>"
            let inApp = !module.contains("UIKit")
                && !module.contains("Foundation")
                && !module.contains("CoreFoundation")
                && !module.contains("libsystem")
                && !module.contains("libobjc")
            return [
                "function": function,
                "file": module,
                "line": 0,
                "inApp": inApp,
            ]
        }
    }
}

File: sdk/react-native/android/src/main/java/com/sentori/SentoriCrashHandler.kt
package com.sentori

import android.content.Context
import android.os.Build
import android.util.Base64
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import java.util.UUID

/**
 * Static crash handler — captures Java/Kotlin uncaught exceptions on
 * Android and writes one event-shaped JSON file per crash to
 * <filesDir>/sentori/pending/<uuid>.json. JS drains that directory on
 * next launch via Sentori.drainPending().
 *
 * What this does NOT do (Phase 7 v0.1):
 *   - native crashes (NDK / SIGSEGV) — Phase 7 explicitly skips signal-
 *     based handlers per ROADMAP.
 *   - ANR detection — deferred to v0.2.
 */
object SentoriCrashHandler {

    private const val PREFS = "sentori"
    private const val PENDING_DIR_NAME = "sentori/pending"

    @Volatile private var appCtx: Context? = null
    @Volatile private var previousHandler: Thread.UncaughtExceptionHandler? = null

    @JvmStatic
    fun register(context: Context) {
        appCtx = context.applicationContext
        previousHandler = Thread.getDefaultUncaughtExceptionHandler()
        Thread.setDefaultUncaughtExceptionHandler { thread, throwable ->
            try {
                write(throwable)
            } catch (_: Throwable) {
                // never throw inside the crash handler
            }
            previousHandler?.uncaughtException(thread, throwable)
        }
        // Phase 42 sub-F.01: have the screenshot helper track the
        // foreground Activity so it knows which Window to PixelCopy.
        (appCtx as? android.app.Application)?.let {
            SentoriScreenshotCapture.register(it)
            // v0.9.6 #2 — replay capture also tracks the activity for
            // wireframe view-tree walks.
            SentoriReplayCapture.register(it)
        }
    }

    @JvmStatic
    fun setConfig(config: Map<String, Any?>) {
        val ctx = appCtx ?: return
        val prefs = ctx.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        val edit = prefs.edit()
        edit.clear()
        for ((k, v) in config) {
            when (v) {
                is String -> edit.putString(k, v)
                is Int -> edit.putInt(k, v)
                is Boolean -> edit.putBoolean(k, v)
                else -> {}
            }
        }
        edit.apply()
    }

    @JvmStatic
    fun consumePending(): List<String> {
        val dir = pendingDir() ?: return emptyList()
        if (!dir.exists()) return emptyList()
        val out = mutableListOf<String>()
        val files = dir.listFiles { f -> f.extension == "json" } ?: emptyArray()
        for (f in files) {
            try {
                out.add(f.readText())
            } catch (_: Throwable) {
                // skip unreadable file
            }
            f.delete()
        }
        return out
    }

    // ── internals ────────────────────────────────────────────────

    private fun pendingDir(): File? {
        val ctx = appCtx ?: return null
        val dir = File(ctx.filesDir, PENDING_DIR_NAME)
        if (!dir.exists()) dir.mkdirs()
        return dir
    }

    private fun configMap(): Map<String, String> {
        val ctx = appCtx ?: return emptyMap()
        val prefs = ctx.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        val out = mutableMapOf<String, String>()
        for ((k, v) in prefs.all) if (v is String) out[k] = v
        return out
    }

    private fun write(throwable: Throwable) {
        val cfg = configMap()
        val release = cfg["release"] ?: "unknown"
        val environment = cfg["environment"] ?: "prod"

        val device = JSONObject().apply {
            put("os", "android")
            put("osVersion", Build.VERSION.RELEASE)
            put("model", "${Build.MANUFACTURER} ${Build.MODEL}")
        }

        val app = JSONObject().apply {
            put("version", appVersion())
            put("build", appBuild())
        }

        val error = errorToJson(throwable)

        // v0.9.5 #7 — detect crashes originating from native code (JNI
        // / .so libs). Pure SIGSEGV in a stripped .so won't reach us
        // without breakpad (queued for v1.1), but throws that surface
        // as UnsatisfiedLinkError or have native frames in the stack
        // are tagged so the dashboard can split them out.
        val tags = JSONObject()
        if (SentoriNativeOrigin.looksNative(throwable)) {
            tags.put("native_signal", "true")
        }

        val event = JSONObject().apply {
            put("id", uuidLower())
            put("timestamp", iso8601Now())
            put("kind", "error")
            put("platform", "android")
            put("release", release)
            put("environment", environment)
            put("device", device)
            put("app", app)
            put("user", JSONObject.NULL)
            put("tags", tags)
            put("breadcrumbs", JSONArray())
            put("error", error)
            put("fingerprint", JSONArray())
            put("traceId", JSONObject.NULL)
            put("spanId", JSONObject.NULL)
        }

        // Phase 42 sub-F.05/08: capture screen + view tree before the
        // app dies, attach as `_pendingAttachments` for the JS side to
        // upload on next launch (same shape as iOS sub-E).
        attachPending(event)

        val dir = pendingDir() ?: return
        val file = File(dir, "${uuidLower()}.json")
        try {
            file.writeText(event.toString())
        } catch (_: Throwable) {
            // best-effort
        }
    }

    private fun attachPending(event: JSONObject) {
        val snap = try {
            SentoriScreenshotCapture.captureKeyWindow()
        } catch (_: Throwable) {
            null
        } ?: return
        val pending = JSONArray()

        @Suppress("UNCHECKED_CAST")
        val sc = snap["screenshot"] as? Map<String, Any>
        if (sc != null) {
            val b64 = sc["base64"] as? String
            val mt = (sc["mediaType"] as? String) ?: "image/webp"
            if (b64 != null) {
                pending.put(JSONObject().apply {
                    put("kind", "screenshot")
                    put("base64", b64)
                    put("mediaType", mt)
                    put("source", "android")
                })
            }
        }
        val vt = snap["viewTree"]
        if (vt != null) {
            // Convert Map → JSONObject → string → base64
            val asJson = SentoriScreenshotCapture.toJson(vt)
            val base64 = Base64.encodeToString(
                asJson.toString().toByteArray(Charsets.UTF_8),
                Base64.NO_WRAP,
            )
            pending.put(JSONObject().apply {
                put("kind", "viewTree")
                put("base64", base64)
                put("mediaType", "application/json")
                put("source", "android")
            })
        }
        if (pending.length() > 0) {
            event.put("_pendingAttachments", pending)
        }
    }

    private fun errorToJson(throwable: Throwable): JSONObject {
        return JSONObject().apply {
            put("type", throwable.javaClass.name)
            put("message", throwable.message ?: "")
            put("stack", framesToJson(throwable))
            val cause = throwable.cause
            if (cause != null && cause !== throwable) {
                put("cause", errorToJson(cause))
            } else {
                put("cause", JSONObject.NULL)
            }
        }
    }

    private fun framesToJson(throwable: Throwable): JSONArray {
        val arr = JSONArray()
        for (f in throwable.stackTrace) {
            val frame = JSONObject().apply {
                put("function", "${f.className}.${f.methodName}")
                put("file", f.fileName ?: "<unknown>")
                put("line", f.lineNumber.coerceAtLeast(0))
                put("inApp", isInApp(f))
            }
            arr.put(frame)
        }
        return arr
    }

    private fun isInApp(f: StackTraceElement): Boolean {
        val cls = f.className
        if (cls.startsWith("android.")) return false
        if (cls.startsWith("androidx.")) return false
        if (cls.startsWith("java.")) return false
        if (cls.startsWith("javax.")) return false
        if (cls.startsWith("kotlin.")) return false
        if (cls.startsWith("kotlinx.")) return false
        if (cls.startsWith("com.facebook.react.")) return false
        if (cls.startsWith("com.android.")) return false
        if (cls.startsWith("dalvik.")) return false
        if (cls.startsWith("sun.")) return false
        return true
    }

    private fun iso8601Now(): String {
        val f = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US)
        f.timeZone = TimeZone.getTimeZone("UTC")
        return f.format(Date())
    }

    private fun uuidLower(): String = UUID.randomUUID().toString().lowercase(Locale.US)

    private fun appVersion(): String {
        val ctx = appCtx ?: return "0.0.0"
        return try {
            val pi = ctx.packageManager.getPackageInfo(ctx.packageName, 0)
            pi.versionName ?: "0.0.0"
        } catch (_: Throwable) {
            "0.0.0"
        }
    }

    private fun appBuild(): String {
        val ctx = appCtx ?: return "0"
        return try {
            val pi = ctx.packageManager.getPackageInfo(ctx.packageName, 0)
            pi.longVersionCode.toString()
        } catch (_: Throwable) {
            "0"
        }
    }
}
